libraries/ms-common/inc/ms_test_helper_datagram.h

```diff
@@ -0,0 +1,29 @@
+#pragma once
+/* Test helpers for datagrams
+ */
+
+#include "bootloader_can.h"
+#include "can_datagram.h"
+#include "ms_test_helpers.h"
+#include "status.h"
+
+#define DATAGRAM_PROCESS_ALL(e)                                 \
+  while (can_datagram_get_status() == DATAGRAM_STATUS_ACTIVE) { \
+    MS_TEST_HELPER_AWAIT_EVENT(e);                              \
+    can_datagram_process_event(&e);                             \
+    can_process_event(&e);                                      \
+  }
+
+// initialize the datagram helper and dependencies
+void init_datagram_helper(CanStorage *can_storage, CanSettings *can_settings, uint8_t board_id,
+                          CanDatagramSettings *can_datagram_settings);
+
+// mock send the datagram |tx_config| from the client script, with datagram id of 0
+// this function only process events until all can messages are sent
+// which may leave some datagram rx events unprocessed,
+// use DATAGRAM_PROCESS_ALL to finish processing all events
```

Comment:
I notice you don't always call `DATAGRAM_PROCESS_ALL` in the tests after calling `mock_tx_datagram` - when exactly does it need to be called? Would it be better to just call it at the end of `mock_tx_datagram`?

Reply:
if the `rx_cmpl_cb` from the mock datagram that was received starts another datagram transmission, the datagram never enters the `RX_COMPLETE` state, so the `DATAGRAM_PROCESS_ALL` will process the incoming transmission from the board as well. So it can't be called if the response should be received by the test helper